basil/HL/tektronix_oscilloscope.py

```diff
@@ -0,0 +1,197 @@
+        if source == "NONE":
+            display_on = False
+        else:
+            display_on = bool(
+                int(self._intf.query("display:global:{}:state?".format(source)))
+            )
+        return display_on
+
+    def _setup_curve_query(self, parameters, channel=1):
+        """Setup the instrument for the curve query operation"""
+
+        # extract the job parameters
+        wave_type, channel, encoding, bit_nr, datatype, rec_len = parameters['CH' + str(channel)]
+
+        # Switch to the source and setup the data encoding
+        self._intf.write("data:source {}".format('CH' + channel))
+        self.set_data_encoding('ascii')
+        self._intf.write("WFMOUTPRE:BIT_NR {}".format(bit_nr))
+
+        # Set the start and stop point of the record
+        rec_len = self._intf.query("horizontal:recordlength?").strip()
+        self._intf.write("data:start 1")
+        self._intf.write("data:stop {}".format(rec_len))
+        return wave_type
+
+    def _post_process_analog(self, source_data, x_scale, channel=1):
+        """Post processes analog channel data"""
+
+        # Normal analog channels must have the vertical scale and offset applied
+        offset = float(self._intf.query("WFMOutpre:YZEro?"))
+        scale = float(self._intf.query("WFMOutpre:YMUlt?"))
+        source_data = [scale * i + offset for i in source_data]
+
+        # Include y-scale information with analog channel waveforms
+        y_scale = self._get_yscale(channel=channel)
+
+        return channel, source_data, x_scale, y_scale
+
+    def get_waveform(self, channel=1):
+        """Returns an iterator that yields the source data from the oscilloscope"""
+
+        jobs = self._make_jobs(channel=channel)
+
+        # remember the state of the acquisition system and then stop acquiring waveforms
+        acq_state = self._intf.query("ACQuire:STATE?").strip()
+        self._intf.write("ACQuire:STATE STOP")
+        wave_type = self._setup_curve_query(jobs, channel=channel)
+
+        # Horizontal scale information
+        x_scale = self._get_xscale()
+        if x_scale is not None:
+            source_data = [int(i) for i in self.get_data(channel=channel).replace('\n', '').split(',')]
+            if wave_type is WaveType.DIGITAL:
+                raise Exception(
+                    "Data Type Digital analysis not yet implemented"
+                )
+
+            elif wave_type is WaveType.ANALOG:
+                ret_val = self._post_process_analog(source_data, x_scale, channel=channel)
+
+            elif wave_type is WaveType.MATH:
+                raise Exception(
+                    "Data Type Math analysis not yet implemented"
+                )
+
+            else:
+                raise Exception(
+                    "It should have been impossible to execute this code"
+                )
+
+        # # Restore the acquisition state
+        self._intf.write("ACQuire:STATE {}".format(acq_state))
+        return ret_val
```

Comment:
This function can be simplified by

    ....
    
    if wave_type is WaveType.ANALOG:
        ret_val = self._post_process_analog(source_data, x_scale, channel=channel)
    else:
        raise NotImplementedError(f"Analysis for type {wave_type} data not yet implemented!")
    
    # Restore the acquisition state
    self._intf.write("ACQuire:STATE {}".format(acq_state))
    return ret_val

Also, I would never write an Exception for code which should not be executed in production code, this is more like a debug strategy ;)